Remora.Commands/Builders/CommandParameterBuilder.cs

```diff
@@ -0,0 +1,479 @@
+//
+//  CommandParameterBuilder.cs
+//
+//  Author:
+//       Jarl Gullberg <jarl.gullberg@gmail.com>
+//
+//  Copyright (c) Jarl Gullberg
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Remora.Commands.Attributes;
+using Remora.Commands.Conditions;
+using Remora.Commands.Extensions;
+using Remora.Commands.Signatures;
+
+namespace Remora.Commands.Builders;
+
+/// <summary>
+/// A builder class for command parameters.
+/// </summary>
+public class CommandParameterBuilder
+{
+    private readonly int _index;
+    private readonly CommandBuilder _builder;
+    private readonly List<Attribute> _attributes;
+    private readonly List<ConditionAttribute> _conditions;
+
+    private string _name;
+    private bool _isGreedy;
+    private bool _isOptional;
+    private string? _description;
+    private object? _defaultValue;
+    private Type? _parameterType;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="CommandParameterBuilder"/> class.
+    /// </summary>
+    /// <param name="builder">The command this parameter belongs to.</param>
+    /// <param name="type">The optional type of the parameter.</param>
+    /// <remarks>If <paramref name="type"/> is null, <see cref="WithType"/> MUST be called before <see cref="Finish"/>.</remarks>
+    public CommandParameterBuilder(CommandBuilder builder, Type? type)
+    {
+        _index = builder.Parameters.Count;
```

Comment:
Storing the index like this might be somewhat troublesome. I'm also not keen on the constructor modifying the enclosing builder - constructors with side effects are hard to reason about.

Reply:
I could take it as a constructor argument (or perhaps as a method), but both of those are prone to misuse I fear; also seeing as the only place you can use parameters is the command builder anyway, I figured this was probably OK.

Reply:
Coming back to this PR now to finally finish it (did I really open this in 2022?); I've come to realize that the entire reason that this index was here to begin with was to ensure that parameters were put in the right indices, however as it's an immutable list based on the position that the command was built in, this shouldn't be necessary.